### PR TITLE
Openbrush follow-up grant application 

### PR DESCRIPTION
**Pull request by themarkian**
### Project Abstract

The mission of this project is to make ink! usable and facilitate WASM ecosystem adoption. 

To be successful with this mission, we have outlined several steps that would need to be taken.
1. OpenBrush is a library for secure smart contract development on ink!, same as
   [OpenZeppelin](https://github.com/OpenZeppelin/openzeppelin-contracts) on Ethereum.
2. OpenBrush should serve as a starting point for every developer entering a Polkadot(ink!) ecosystem.
3. In addition to having a standard implementation, Openbrush should also have macroses to simplify and abstract up applications(smart contracts) development.
4. For Openbrush to fulfill its purpose, several changes must be made & several blockers to be overcome in `ink!` and in Substrate's `pallet-contracts`. We are committed to doing that in the context of this project.

It is an application for a follow-up grant. The [initial grant](https://github.com/w3f/Grants-Program/pull/497) covered the first and second milestones:
 - [Openbrush Milestone 1 delivery report](https://github.com/w3f/Grant-Milestone-Delivery/pull/249)
 - [Openbrush Milestone 2 delivery report](https://github.com/w3f/Grant-Milestone-Delivery/pull/261)

This grant aims to cover milestones 3-6.

### For which grant [level](https://github.com/w3f/Grants-Program#level_slider-levels) are you applying? 
  - [ ] **Level 1**:  Up to $10,000, 2 approvals
  - [x] **Level 2**:  Up to $50,000, 3 approvals
  - [ ] **Level 3**:  Unlimited, 5 approvals (for > $100k Web3 Foundation Council approval)

### Application Checklist

- [x] The [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) has been copied, renamed ( `project_name.md`) and updated.
- [x] A BTC or Ethereum (DAI/USDT) address for the payment of the milestones is provided inside the application.
- [x] I have read and acknowledged the [terms and conditions](https://github.com/w3f/Grants-Program/blob/master/docs/T&Cs.md).
- [x] The software delivered for this grant will be released under an open-source license specified in the application.
- [x] The initial PR contains only one commit (squash and force-push if needed).
- [x] The grant will only be announced once the first milestone [has been accepted](https://github.com/w3f/Grant-Milestone-Delivery#process).

### CODE REVIEW

**Comment by semuelle** [comment]
Thanks for the application. I'm happy to see OpenBrush moving forward. We will look into it as soon as possible.

**Comment by mmagician** [comment]
@cmichi Please have a look at your convenience.

File: applications/openbrush-follow-up.md [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+* **FTE:**  2.5
+* **Costs:** 35,000 USD
+
+| Number | Deliverable                      | Specification                                                                                                                             |
+| -----  | -----------                      | -------------                                                                                                                             |
+| 0a.    | License                          | MIT                                                                                                                                       |
+| 0b.    | Documentation                    | We will provide inline documentation, example of upgradable contracts. |
+| 0c.    | Testing Guide                    | We will add tests to cover upgradability of contracts.                                                           |
+| 1.     | Implement delegated call         | We will find and provide the idea of how [delegeted call](https://docs.soliditylang.org/en/v0.4.21/introduction-to-smart-contracts.html#delegatecall-callcode-and-libraries) can be implemented in `contract-pallet`. Help with it's implementation.            |
```
Comment (athei) [inline applications/openbrush-follow-up.md:264]: Why is the `delegatecall` needed for upgrade ability? AFAIK it is only needed for library contracts which are not relevant here. Correct me if I am wrong. I was under the impression that https://github.com/paritytech/substrate/pull/8909 implemented everything needed for upgradeable contracts.

Comment (xgreenx) [reply applications/openbrush-follow-up.md:264]: The idea of an upgradeable contract in Solidity, that you will execute the code of another contract in the context of the current contract(you will affect the storage of the current contract by the code of callee contract). 

This change doesn't allow that. It allows only to call another without the ability to return back by the `TAIL_CALL` flag.

Comment (athei) [reply applications/openbrush-follow-up.md:264]: I know what delegatecall is. I just didn't know it was used for upgradeability. But it makes sense I guess. But you need to make sure that the new version does not change the storage layout, right?

Comment (xgreenx) [reply applications/openbrush-follow-up.md:264]: Yes, the developer of the upgradable contract must worry about layout. We will create a list of rules that the developer must follow in order not to break the layout. [OpenZeppelin](https://docs.openzeppelin.com/upgrades-plugins/1.x/writing-upgradeable) did the same for solidity. Our library will provide standards in upgradable form too.

**Comment by mmagician** [comment]
I'm not 100% sure at this stage, I guess we'll find out once we get there :)
If you're happy to go ahead, then let me know once you've updated the scope of the application and I'll share with the rest of the grants team - but given that it has a go-ahead from ink! team it should be straightforward.

**Comment by semuelle** [comment]
Thanks for the update, @xgreenx. Two more questions: 

- What is the difference between the existing PSP-22 and deliverable 4.1 (`Create Proposal for Fungible token`)?
- Could you list the (possible) extensions mentioned in deliverable 4.2? I assume you mean `Mintable`, `Pausable`, etc.?

**Review by Noc2 — APPROVED** [review]

**Comment by semuelle** [comment]
Hey @0xMarkian & @xgreenx, can I ask what the status of M5 and related issues is? Is this still on the roadmap? I remember vaguely seeing it discussed somewhere on here, but can't find it anymore.

**Comment by xgreenx** [comment]
I also can't find where we discuss it=) Maybe in the old element.
Milestone 5 is ready. It seems we forgot to deliver it.
- Contribute to ink! with fixing of events- ink! teams decided to fix it internally without our help we only participated in the [issue](https://github.com/paritytech/ink/issues/809#issuecomment-968036080).
- Re-work the storage of contracts - Is done and [merged](https://github.com/paritytech/ink/issues/1134).
- Refactor of implementation according changes in ink! - Also is done via [a bunch of PRs](https://github.com/Brushfam/openbrush-contracts/pulls?q=is%3Apr+is%3Aclosed+ink+4).

I think we need to create a delivery report.

**Comment by semuelle** [comment]
> I think we need to create a delivery report.

That would be great. Thanks!

**Comment by keeganquigley** [comment]
Thanks @xgreenx does your team still plan on submitting a delivery report?

**Comment by Artemka374** [comment]
Hey! This milestone was created as a part of Supercolony's work, so we are not sure that we can deliver it as a part of Brushfam. Is it a problem, and if it is, does there exist some solution to it?

**Comment by semuelle** [comment]
> Is it a problem, and if it is, does there exist some solution to it?

That is indeed not so trivial. I'll check with our legal team if we can assign the grant to you. Otherwise, we might have to cancel the old one and sign a new one.

**Comment by 0xMarkian** [comment]
Hey @semuelle, thanks a lot for making the inquiry! However, I believe Brushfam cannot receive compensation for this milestone as it wasn't executed by us, but rather by SuperColony. 
That said, we value our relationship and if it will be helpful for w3f Brushfam can do delivery itself.

**Comment by semuelle** [comment]
Thanks for the quick reply, @0xMarkian. My question was specifically for the purpose of transferring the grant from SuperColony to BrushFam so that you could deliver and get compensated for it. It would, however, require SuperColony's written approval. 

If that fails, of course we'd be happy to review your delivery of M5. I understand that it isn't your responsibility, though, and would be voluntary. I can see that the work was done, so I wouldn't expect you to put more work into it now.

**Comment by 0xMarkian** [comment]
Hey, thanks for understanding. Let us know if M5 delivery will be of help for w3f. If so we will do it!

**Comment by semuelle** [comment]
From my point of view, OpenBrush is doesn't need our stamp of approval anymore. If you're not reliant on it, we'd simply cancel the last milestone and call it a day.

**Comment by 0xMarkian** [comment]
Hey, apologies, missed this message. Yes, sounds good to me!

**Comment by semuelle** [comment]
Thanks for the straightforward resolution, @0xMarkian. The remaining milestone is officially cancelled. 
Looking forward to your next application.

**Comment by 0xMarkian** [comment]
Thank you @semuelle !

**Pull request merged**